File: algorithm_practice_python/leetcode/arrays101/SquaresOfaSortedArray.py
from typing import List


class SquaresOfaSortedArray:
    # First try
    # def sortedSquares(self, A: List[int]) -> List[int]:
    #     return sorted([x * x for x in A])

    def sortedSquares(self, A: List[int]) -> List[int]:
        N = len(A)
        j = 0
        while j < N and A[j] <= 0:
            j += 1

        i = j - 1
        ans = []

        while i >= 0 and j < N:
            if A[i] ** 2 < A[j] ** 2:
                ans.append(A[i] ** 2)
                i -= 1
            else:
                ans.append(A[j] ** 2)
                j += 1

        while i >= 0:
            ans.append(A[i] ** 2)
            i -= 1

        while j < N:
            ans.append(A[j] ** 2)
            j += 1

        return ans

File: algorithm_practice_java/src/main/java/leetcode/arrays101/SquaresOfaSortedArray.java
package leetcode.arrays101;

/**
 * url: https://leetcode.com/explore/learn/card/fun-with-arrays/511/in-place-operations/3261/
 */
public class SquaresOfaSortedArray {
    //simple but time : NlogN / space : N
/*    public static int[] sortedSquares(int[] A) {
        for (int i = 0; i < A.length; i++) {
            A[i] = A[i] * A[i];
        }
        Arrays.sort(A);
        return A;
    }*/

    //two pointers. time : N / space : N
    public static int[] sortedSquares(int[] A) {
        int N = A.length;
        int j = 0;
        while (j < N && A[j] < 0) {
            j++;
        }
        int i = j - 1;
        int[] ans = new int[N];
        int t = 0;

        while (i >= 0 && j < N) {
            if (A[i] * A[i] < A[j] * A[j]) {
                ans[t++] = A[i] * A[i];
                i--;
            } else {
                ans[t++] = A[j] * A[j];
                j++;
            }
        }

        while (i >= 0) {
            ans[t++] = A[i] * A[i];
            i--;
        }

        while (j < N) {
            ans[t++] = A[j] * A[j];
            j++;
        }
        return ans;
    }
}
